### PR TITLE
Unify RelayChainInterface error handling and introduce async

### PR DESCRIPTION
**Pull request by skunert**
This PR includes more changes in preparation of network communication with the relay chain:
- Make `RelayChainInterface` methods async
- Make methods return results (Over network, all operations could fail)
- Logging of errors is moved out of the interface to the caller side
- Unify error handling by introducing `RelayChainError` with enum variants for different errors
- Introduce `new_best_notification_stream` to the interface. Over RPC, we only receive the header instead of `BlockImportNotification`. In some places we were checking if a new block is the new best. These places now consume `new_best_notification_stream`.

This PR is not expected to change behaviour.

### CODE REVIEW

**Review by bkchr — COMMENTED** [review]
Sorry for the delay.

Looks good, but can still be a little bit improved here and there.

File: client/service/src/lib.rs [hunk 1]
```diff
@@ -107,10 +107,13 @@ where
 		.spawn_essential_handle()
 		.spawn("cumulus-consensus", None, consensus);
 
+	let overseer_handle = relay_chain_interface
+		.overseer_handle()
+		.map_err(|e| sc_service::Error::Other(e.to_string()))?
```
Comment (bkchr) [inline client/service/src/lib.rs:112]:
```suggestion
		.map_err(|e| sc_service::Error::Application(Box::new(e)))?
```

File: client/service/src/lib.rs [hunk 2]
```diff
@@ -192,10 +193,13 @@ where
...
```
Comment (bkchr) [inline client/service/src/lib.rs:198]:
```suggestion
		.map_err(|e| sc_service::Error::Application(Box::new(e)))?
```

File: client/relay-chain-local/src/lib.rs [hunk 3]
```diff
@@ -303,7 +273,7 @@ where
 	let block_id = BlockId::Hash(hash);
 	match backend.blockchain().status(block_id) {
 		Ok(BlockStatus::InChain) => return Ok(BlockCheckStatus::InChain),
```
Comment (bkchr) [inline client/relay-chain-local/src/lib.rs:275]:
```suggestion
	match backend.blockchain().status(block_id)? {
		BlockStatus::InChain => return Ok(BlockCheckStatus::InChain),
```

File: client/relay-chain-local/src/lib.rs [hunk 4]
```diff
@@ -303,7 +273,7 @@ where
...
-		Err(err) => return Err(WaitError::BlockchainError(hash, err)),
+		Err(err) => return Err(RelayChainError::BlockchainError(err)),
```
Comment (bkchr) [inline client/relay-chain-local/src/lib.rs:276]:
```suggestion
```

File: client/network/src/lib.rs [hunk 5]
```diff
@@ -222,6 +218,7 @@ impl TryFrom<&'_ CollationSecondedSignal> for BlockAnnounceData {
 /// chain. If it is at the tip, it is required to provide a justification or otherwise we reject
 /// it. However, if the announcement is for a block below the tip the announcement is accepted
 /// as it probably comes from a node that is currently syncing the chain.
+#[derive(Clone)]
```
Comment (bkchr) [inline client/network/src/lib.rs:221]: For what do we need the clone?

File: client/pov-recovery/src/lib.rs [hunk 6]
```diff
@@ -435,28 +442,40 @@ where
...
+				client_for_closure.session_index_for_child(&block_id).await.map_err(|e| {
+					tracing::error!(
+						target: LOG_TARGET,
+						error = ?e,
+						"Failed to fetch session index.",
+					)
+				});
+
+			session_index_result
+				.and_then(|v| Ok(pending_availability_result?.map(|pa| (pa, v))))
```
Comment (bkchr) [inline client/pov-recovery/src/lib.rs:475]:
```suggestion
				.and_then(|v| pending_availability_result.map(|pa| (pa, v)))
```

File: client/relay-chain-interface/Cargo.toml [hunk 7]
```diff
@@ -15,7 +15,10 @@ sp-runtime = { git = "https://github.com/paritytech/substrate", branch = "master
 sp-blockchain = { git = "https://github.com/paritytech/substrate", branch = "master" }
 sp-state-machine = { git = "https://github.com/paritytech/substrate", branch = "master" }
 sc-client-api = { git = "https://github.com/paritytech/substrate", branch = "master" }
+sc-service = { git = "https://github.com/paritytech/substrate", branch = "master" }
 
+futures = { version = "0.3.1", features = ["compat"] }
```
Comment (bkchr) [inline client/relay-chain-interface/Cargo.toml:20]:
```suggestion
futures = "0.3.1"
```

**Review by bkchr — APPROVED** [review]

File: client/network/src/lib.rs (hunk 5 above)
Comment (skunert) [reply client/network/src/lib.rs:221]: I was having lifetime issues while moving function calls to the async block in the validate method: https://github.com/paritytech/cumulus/pull/909/files#diff-0fe3ce0d7087ba7f979d9065d8d812b23562d2f4ed5341719aa8480dc48c74cdR345

File: client/pov-recovery/src/lib.rs (hunk 6 above)
Comment (skunert) [reply client/pov-recovery/src/lib.rs]: This suggestion does not work here as it changes the return type. However, I tried to improve the readability of this section

File: client/network/src/lib.rs (hunk 5 above)
Comment (skunert) [reply client/network/src/lib.rs:221]: Done

**Pull request merged**